support-frontend/app/utils/CheckoutValidationRules.scala

```diff
@@ -123,29 +124,175 @@ object DigitalPackValidation {
-
   def passes(createSupportWorkersRequest: CreateSupportWorkersRequest): Boolean = {
 
-    val hasDeliveryAddressInUKAndPaidInGbp = createSupportWorkersRequest.deliveryAddress match {
-      case Some(address) => deliveredToUkAndPaidInGbp(address.country, createSupportWorkersRequest.product.currency)
-      case None => false
+    val maybeValid = for {
+      address <- createSupportWorkersRequest.deliveryAddress
+    } yield {
+
+      val deliveryAddressHasAddressLine1AndCity = hasAddressLine1AndCity(address)
+
+      PaidProductValidation.passes(createSupportWorkersRequest) &&
+        createSupportWorkersRequest.firstDeliveryDate.nonEmpty &&
+        hasAddressLine1AndCity(createSupportWorkersRequest.billingAddress) &&
+        deliveryAddressHasAddressLine1AndCity
+
     }
 
-    val deliveryAddressHasAddressLine1AndCity = createSupportWorkersRequest.deliveryAddress match {
-      case Some(address) => hasAddressLine1AndCity(address)
-      case None => false
+    maybeValid.contains(true)
+
+  }
+
+}
+
+object PaperValidation {
+
+  import AddressAndCurrencyValidationRules._
+
+  def passes(createSupportWorkersRequest: CreateSupportWorkersRequest, fulfilmentOptions: FulfilmentOptions): Boolean = {
+
+    val maybeValid = for {
+      address <- createSupportWorkersRequest.deliveryAddress
+      postCode <- address.postCode
+    } yield {
+
+      val hasDeliveryAddressInUKAndPaidInGbp = deliveredToUkAndPaidInGbp(address.country, createSupportWorkersRequest.product.currency)
+      val deliveryAddressHasAddressLine1AndCity = hasAddressLine1AndCity(address)
+      val validPostcode = fulfilmentOptions match {
+        case HomeDelivery => postcodeIsWithinDeliveryArea(postCode)
+        case Collection => true
+        case Domestic => false
+        case RestOfWorld => false
+        case NoFulfilmentOptions => false
+      }
+
+      PaidProductValidation.passes(createSupportWorkersRequest) &&
+        hasDeliveryAddressInUKAndPaidInGbp &&
+        createSupportWorkersRequest.firstDeliveryDate.nonEmpty &&
+        hasAddressLine1AndCity(createSupportWorkersRequest.billingAddress) &&
+        deliveryAddressHasAddressLine1AndCity &&
+        validPostcode
+
     }
 
-    PaidProductValidation.passes(createSupportWorkersRequest) &&
-    hasDeliveryAddressInUKAndPaidInGbp &&
-    hasFirstDeliveryDate(createSupportWorkersRequest.firstDeliveryDate) &&
-    hasAddressLine1AndCity(createSupportWorkersRequest.billingAddress) &&
-    deliveryAddressHasAddressLine1AndCity
+    maybeValid.contains(true)
+
+  }
+
+  def getPrefix(postcode: String): String = {
+
+    val formattedPostcode = postcode.replaceAll(" ", "").toUpperCase
+    formattedPostcode.slice(0, formattedPostcode.length - 3)
 
   }
+
+  def postcodeIsWithinDeliveryArea(postcode: String): Boolean =
+    M25_POSTCODE_PREFIXES.contains(getPrefix(postcode))
+
+  // NOTE also in deliveryCheck.js
+
+  val M25_POSTCODE_OLD_PREFIXES = List(
```

Comment:
Is this list duplicated in the client side now? It would be good to pass it through from here to the client so we don't have to maintain two lists.

Reply:
I have done that now